### PR TITLE
Support Roberta embedding models

### LINKED ISSUE #9847: [New Model]: BAAI/bge-m3
### The model to consider.

https://huggingface.co/BAAI/bge-m3

### The closest model vllm already supports.

_No response_

### What's your difficulty of supporting the model you want?

_No response_

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by maxdebayser**
This PR adds support for Roberta embedding models. It's mostly the same as the Bert architecture, the only thing that changes is the padding token in the Embedding layer so this PR tries to reuse Bert modeling classes as much as possible. For some of the models we also need head size 32, so this size is added to the kernels here.

cc: @robertgshaw2-neuralmagic , @DarkLight1337 

FIX #9847

### CODE REVIEW

**Comment by DarkLight1337** [comment]
Also, remember to update the Supported Models page.

**Comment by DarkLight1337** [comment]
It would also be nice to consider adding the `SequenceClassification` variant of the model (to solve #8022)

**Comment by rpvelloso** [comment]
great! What's the ETA on this???

**Comment by maxdebayser** [comment]
> Model implementation looks good (though a bit hacky). Can you add tests for the model?

Sure, I'll add the tests. I don't disagree that this is a bit hacky. Should we make the Bert classes more generic so that we can pass the embedding layer class as a parameter?

**Comment by DarkLight1337** [comment]
> > Model implementation looks good (though a bit hacky). Can you add tests for the model?
> 
> Sure, I'll add the tests. I don't disagree that this is a bit hacky. Should we make the Bert classes more generic so that we can pass the embedding layer class as a parameter?

That would be great. Another way would be to have an abstract `_init_embedding` etc. so subclasses can decide how to initialize each submodule.

File: vllm/model_executor/models/roberta.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+   Attributes:
+       model: An instance of RobertaModel used for forward operations.
+       _pooler: An instance of Pooler used for pooling operations.
+   """
+
+    def __init__(self,
+                 config: RobertaConfig,
+                 cache_config: Optional[CacheConfig] = None,
+                 quant_config: Optional[QuantizationConfig] = None,
+                 pooler_config: Optional[PoolerConfig] = None) -> None:
```
Comment (DarkLight1337) [inline vllm/model_executor/models/roberta.py:72]: We should now use `VllmConfig` since #9983 . Also, please refactor the inner models as well.

**Comment by DarkLight1337** [comment]
Please fix the formatting errors.

**Comment by DarkLight1337** [comment]
Also we have to merge from main to fix the CI failures.

**Review by DarkLight1337 — APPROVED** [review]
LGTM, thanks for adding this!

**Comment by maxdebayser** [comment]
@DarkLight1337, I realized that with Roberta models the position_ids start at padding_idx + 1 (see [here](https://github.com/huggingface/transformers/blob/a3d69a8994d673899608a7c17fbf4f953f50474e/src/transformers/models/roberta/modeling_roberta.py#L133) and [here](https://github.com/huggingface/transformers/blob/a3d69a8994d673899608a7c17fbf4f953f50474e/src/transformers/models/roberta/modeling_roberta.py#L1669) )

I've added a line of code to increment all position_ids by that amount. Without this, the results I get in the STS12 task from the MTEB benchmark for `intfloat/multilingual-e5-large` is this, 0.53, which is way off. With the change I get  0.80, which is correct. 

In my tests, all the position_ids in vllm for the embedding use case start with 0 and end with len()-1 and there are no padding tokens because the input tensors are 1-dimensional without padding. For example:

```
input_ids=tensor([     0,   1284,     70,   1821,   6275,    111,  21455,      6,      4,
            70,  21640,  31486,    111,     70,  15437,    509,     10,  63819,
             6,      5,      2,      0,   2367,     83,   1286,      6,      4,
            70,  16648,    111,  29700,    621,    959, 133888,     47, 137447,
          2363, 102880,    111,  77021,     47,  16839,  27289,      6,      5,
             2], device='cuda:0')
position_ids=tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
        18, 19, 20,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14,
        15, 16, 17, 18, 19, 20, 21, 22, 23, 24], device='cuda:0')
```

Is there a scenario in which there could be a presence of padding tokens? (Except for the case in which the user inserts <pad> in the input text).

**Comment by DarkLight1337** [comment]
> I've added a line of code to increment all position_ids by that amount. Without this, the results I get in the STS12 task from the MTEB benchmark for intfloat/multilingual-e5-large is this, 0.53, which is way off. With the change I get 0.80, which is correct.

Hmm, we may need to add a correctness test that compares against HF then.

> Is there a scenario in which there could be a presence of padding tokens? (Except for the case in which the user inserts in the input text).

Don't think so, since vLLM encodes each prompt separately. Just to be sure, you can add an assertion statement so we know if our assumption is false.

**Comment by sorenmc** [comment]
> It would also be nice to consider adding the `SequenceClassification` variant of the model (to solve #8022)

Are you still considering adding this to the pr? If not i could try to make an attempt.

**Comment by maxdebayser** [comment]
@DarkLight1337 , I've added an `assert` on the `position_ids`.

> Hmm, we may need to add a correctness test that compares against HF then.

Yes, but it would have to be `sentence-transformers`. In the transformers library the pooled output is obtained by running the last hidden states through the `pooler` layer. But in sentence-transformers, this output is discarded and the hidden states are pooled using MEAN, CLS ... like we do and then normalized. Would it be OK to add this dependency? If yes, can we do this in another PR?

**Comment by DarkLight1337** [comment]
The existing embedding models tests for text-only models already use sentence-transformers, so it should be pretty straightforward to add this model to the list.

**Comment by maxdebayser** [comment]
The discussion on `SequenceClassification` models continues here in this other PR: https://github.com/vllm-project/vllm/pull/10400

**Pull request merged**